lib/ansible/modules/system/listen_ports_facts.py

```diff
@@ -0,0 +1,353 @@
+          type: int
+          sample: 609
+        port:
+          description: The port the server is listening on.
+          returned: always
+          type: int
+          sample: 514
+        protocol:
+          description: The network protocol of the server.
+          returned: always
+          type: str
+          sample: "udp"
+        stime:
+          description: The start time of the listening process.
+          returned: always
+          type: str
+          sample: "Thu Feb  2 13:29:45 2017"
+        user:
+          description: The user who is running the listening process.
+          returned: always
+          type: str
+          sample: "root"
+'''
+
+import re
+import platform
+from ansible.module_utils._text import to_native
+from ansible.module_utils.basic import AnsibleModule
+
+
+def netStatParse(raw):
+    results = list()
+    for line in raw.splitlines():
+        listening_search = re.search('[^ ]+:[0-9]+', line)
+        if listening_search:
+            splitted = line.split()
+            conns = re.search('([^ ]+):([0-9]+)', splitted[3])
+            pidstr = ''
+            if 'tcp' in splitted[0]:
+                protocol = 'tcp'
+                pidstr = splitted[6]
+            elif 'udp' in splitted[0]:
+                protocol = 'udp'
+                pidstr = splitted[5]
+            pids = re.search('(([0-9]+)/(.*)|-)', pidstr)
+            if conns and pids:
+                address = conns.group(1)
+                port = conns.group(2)
+                if (pids.group(2)):
+                    pid = pids.group(2)
+                else:
+                    pid = 0
+                if (pids.group(3)):
+                    name = pids.group(3)
+                else:
+                    name = ''
+                result = dict(pid=int(pid), address=address, port=int(port), protocol=protocol, name=name)
+                if result not in results:
+                    results.append(result)
+            else:
+                raise EnvironmentError('Could not get process information for the listening ports.')
+    return results
+
+
+def applyWhitelist(portspids, whitelist=None):
+    if whitelist is None:
+        whitelist = list()
+    processes = list()
+    for p in portspids:
+        if int(p['port']) not in whitelist and str(p['port']) not in whitelist:
+            process = dict(pid=p['pid'], address=p['address'], port=p['port'], protocol=p['protocol'], name=p['name'], stime=p['stime'], user=p['user'])
+            if process not in processes:
+                processes.append(process)
+    return processes
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            whitelist_tcp=dict(type='list', default=list()),
+            whitelist_udp=dict(type='list', default=list()),
+        ),
+        supports_check_mode=True,
+    )
+
+    if platform.system() != 'Linux':
+        module.fail_json(msg='This module requires Linux.')
+
+    def getPidSTime(pid):
```

Comment:
Not a requirement, just to understand. Why define this as part of the main function and not make this just a global function ? I think it helps with readability.

Reply:
I don't recall a specific reason - agreed it reads better outside the main function.

Reply:
Ah I see now - it relies on the `module` instance created by `AnsibleModule` to use the `run_command()` function. Can `AnsibleModule` be run outside of the main function? Otherwise I can move the other functions inside the main function to make it all more consistent.

Reply:
Ok, I see. What we tend to do is pass the module to the function instead. But honestly, don't bother.